File: 2-ad-hoc/1246.php
<?php
class Car{
	public $PLATE,$START,$END;

	function __construct($ID,$INIT,$END){
		$this->PLATE=$ID;
		$this->START=$INIT;
		$this->END=$END;
	}
	function display(){
		print($this->PLATE." ".$this->START." ".$this->END."|");
	}
}

class Node{
	public $next,$PREV,$VALUE;

	function __construct($CAR_VALUE,$PREV_VALUE){
		$this->value=$CAR_VALUE;
		$this->next=null;
		$this->prev=$PREV_VALUE;
	}
}
class Head_ll{
	public $head,$first,$last;

	function __construct(){
		$this->head=new Node(null,null);
		$this->first=$this->head;
		$this->last=$this->head;
	}
	function insert($new_value){
		if($this->head->value==null){
			$this->last=$this->head;

			$this->head->value=$new_value;
			$this->head->next=new Node(null,$this->head);
			$this->head=$this->head->next;
			return;
		}
		if($this->head->prev==null){
			$this->head->prev=new Node($new_value,null);
			$this->head->prev->next=$this->head;
			$this->head=$this->head->prev;
			$this->first=$this->head;
			return;
		}
		$temp=$this->head;
		$this->head=$this->head->prev;
		$this->head->next=new Node($new_value,$this->head);
		$this->head=$this->head->next;
		$this->head->next=$temp;
		$temp->prev=$this->head;
	}
	function delete(){
		if($this->head->prev==null){
			$this->head->value=null;
			if($this->head->next!=null&&$this->head->next->value!=null){
				$this->head=$this->head->next;
				$this->head->prev=null;
			}
			$this->first=$this->head;
			return;
		}
		if($this->head->value==null){
			$this->head=$this->head->prev;
			$this->head->next=null;
			$this->head->value=null;
			$this->last=$this->head->prev;
			return;
		}
		$temp=$this->head->next;
		$this->head=$this->head->prev;
		$this->head->next=$temp;
		$temp->prev=$this->head;
		if($temp->next==null||$temp->next->value!=null)
			$this->last=$temp->prev;
	}
	function move_start(){
		$this->head=$this->first;
	}
	function move_end(){
		$this->head=$this->last;
	}

	function display_start(){
		$temp=$this->first;
		while($temp->value!=null){
			$temp->value->display();
			$temp=$temp->next;
		}
		print("\n");
	}
	function display_end(){
		$temp=$this->last;
		while($temp!=null&&$temp->value!=null){
			$temp->value->display();
			$temp=$temp->prev;
		}
		print("\n");
	}
	function is_empty(){
		return ($this->first==$this->last&&$this->first->value==null);
	}
}

function let_parking($park,$ID){
	$park->move_start();
	while($park->head->value!=null){
		if($park->head->value->PLATE==$ID){
			$park->delete();
			return;
		}
		$park->head=$park->head->next;
	}
}
function inp_parking($park,$len_park,$ID_car,$len_car){
	$park->move_start();
	if($len_car>$len_park)
		return 0;
	if($park->head->value==null||$park->head->value->START>=$len_car){
		$park->insert(new Car($ID_car,0,$len_car-1));
		return 1;
	}
	$prev_end=$park->head->value->END;
	$park->head=$park->head->next;
	while($park->head->value!=null){
		$curr=$park->head->value;
		if($curr->START-$prev_end>$len_car){
			$park->insert(new Car($ID_car,$prev_end+1,$prev_end+$len_car));
			return 1;
		}
		$prev_end=$curr->END;
		$park->head=$park->head->next;
	}
	if($len_park-$prev_end>$len_car){
		$park->insert(new Car($ID_car,$prev_end+1,$prev_end+$len_car));
		return 1;
	}
	return 0;
}

/*
$TST=new Head_ll();
for($I=0;$I<10;++$I){
	$STR=chr($I%25+65).chr(($I+1)%26+65).chr(($I+2)%26+65);
	$TST->insert(new Car($STR,$I,$I));
}

$TST->move_end();
while(!$TST->is_empty()){
	$TST->head->value->display();
	$TST->delete();
}
echo "\n";
 */

while(($line=readLine())!==false){
	list($length,$quotes)=explode(' ',$line);
	$length=(int)$length;
	$quotes=(int)$quotes;
	$dict=array();
	$parking=new Head_ll();
	$tot=0;
	while($quotes>0){
		$str=explode(' ',readLine());

		if(sizeof($str)<3&&array_key_exists($str[1],$dict)){
			let_parking($parking,$str[1]);
			unset($dict[$str[1]]);
		}else if(sizeof($str)==3&&inp_parking($parking,$length,$str[1],(int)$str[2])){
			$dict[$str[1]]=true;
			$tot+=10;
		}
		//$parking->display_start();

		--$quotes;
	}
	print($tot);
	print("\n");
}
?>

File: 2-ad-hoc/1246.rs
use std::io;
use std::option;

#[derive(Clone)]
struct Car{
    ID:String,
    init:u16,
    end:u16,
}

impl Car{
    pub fn new(identity:String,start:u16,end:u16)->Self{
        Car{ID:identity,init:start,end:end}
    }
    pub fn display_car(&self){
        print!("{} {} {}|",self.ID,self.init,self.end);
    }
}

#[derive(Clone)]
struct Node{
    value:Option<Car>,
    prev:Option<Box<Node>>,
    next:Option<Box<Node>>,
}
impl Node{
    pub fn new(new_car:Option<Car>,prev:Option<Box<Node>>)->Option<Box<Self>>{
        if !prev.is_none(){
            return Some(Box::new(Node{value:new_car,prev:prev,next:None}));
        }
        return Some(Box::new(Node{value:new_car,prev:None,next:None}));
    }
    pub fn get_value(&self)->Option<Car>{
        self.value.clone()
    }
    pub fn change_value(&mut self,new_value:Option<Car>){
        self.value=new_value;
    }
    pub fn display_node(&self){
        if let Some(value)=&self.value{
            value.display_car();
        }
    }
}

struct Head_ll{
    head:Option<Box<Node>>,
    last:Option<Box<Node>>,
    prim:Option<Box<Node>>,
}
impl Head_ll{
    pub fn new()->Self{
        let mut new=Head_ll{head:Node::new(None,None),prim:None,last:None};
        new.prim=new.head.clone();
        new.last=new.head.clone();
        new
    }
    pub fn move_forward(&mut self){
        if let Some(mut forward)=self.head.take(){
            self.head=forward.next.take();
        }
    }
    pub fn get_clone(target:&mut Option<Box<Node>>)->&mut Box<Node>{
        let Some(ref mut res)=target else { todo!() };
        res
    }
    pub fn insert(&mut self,new_car:Car){
        let unpack=Head_ll::get_clone(&mut self.head);
        match unpack.get_value(){
            Some(_)=>{
                println!("There!");
            }None=>{
                unpack.change_value(Some(new_car));
                unpack.next=Node::new(None,Some(unpack.clone()));
                self.last=self.head.clone();
                self.move_forward();
                return;
            }
        }
        match unpack.prev{
            Some(_)=>{}
            None=>{
                println!("NAO");
            }
        }
    }
    pub fn move_start(&mut self){
        self.head=self.prim.clone();
    }
    pub fn move_end(&mut self){
        self.head=self.last.clone();
    }

    pub fn display_start(&self){
        let mut temp=self.prim.as_ref();
        while let Some(current)=temp{
            current.display_node();
            temp=current.next.as_ref();
        }
        print!("\n");
    }
    pub fn display_end(&self){
        let mut temp=self.last.as_ref();
        while let Some(current)=temp{
            current.display_node();
            temp=current.prev.as_ref();
        }
        print!("\n");
    }
}

fn main(){
    let mut tst:Head_ll=Head_ll::new();
    tst.insert(Car::new("1234".to_string(),0,0));
    tst.insert(Car::new("1235".to_string(),0,0));
    tst.display_end();
    tst.display_start();
}
